File: packages/ts/client/src/txbuilder/closePositionOrca.ts
import { type Account, IInstruction, TransactionSigner } from "@solana/kit";
import { findAssociatedTokenPda, Mint, TOKEN_2022_PROGRAM_ADDRESS } from "@solana-program/token-2022";
import { getPositionAddress, WHIRLPOOL_PROGRAM_ADDRESS } from "@orca-so/whirlpools-client";
import { getClosePositionOrcaInstruction, getTunaPositionAddress, TunaPosition } from "../index.ts";

export async function closePositionOrcaInstruction(
  authority: TransactionSigner,
  tunaPosition: Account<TunaPosition>,
  mintA: Account<Mint>,
  mintB: Account<Mint>,
): Promise<IInstruction> {
  const positionMint = tunaPosition.data.positionMint;

  const orcaPositionAddress = (await getPositionAddress(positionMint))[0];

  const tunaPositionAddress = (await getTunaPositionAddress(positionMint))[0];
  const tunaPositionAta = (
    await findAssociatedTokenPda({
      owner: tunaPositionAddress,
      mint: positionMint,
      tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
    })
  )[0];

  const tunaPositionAtaA = (
    await findAssociatedTokenPda({
      owner: tunaPositionAddress,
      mint: mintA.address,
      tokenProgram: mintA.programAddress,
    })
  )[0];

  const tunaPositionAtaB = (
    await findAssociatedTokenPda({
      owner: tunaPositionAddress,
      mint: mintB.address,
      tokenProgram: mintB.programAddress,
    })
  )[0];

  return getClosePositionOrcaInstruction({
    mintA: mintA.address,
    mintB: mintB.address,
    authority,
    tunaPositionMint: positionMint,
    tunaPositionAta,
    tunaPositionAtaA,
    tunaPositionAtaB,
    orcaPosition: orcaPositionAddress,
    tunaPosition: tunaPositionAddress,
    whirlpoolProgram: WHIRLPOOL_PROGRAM_ADDRESS,
    tokenProgramA: mintA.programAddress,
    tokenProgramB: mintB.programAddress,
    token2022Program: TOKEN_2022_PROGRAM_ADDRESS,
  });
}

File: packages/rust/client/src/txbuilder/close_position_orca.rs
use crate::accounts::TunaPosition;
use crate::get_tuna_position_address;
use crate::instructions::ClosePositionOrca;
use orca_whirlpools_client::get_position_address;
use solana_program::instruction::Instruction;
use solana_program::pubkey::Pubkey;
use spl_associated_token_account::get_associated_token_address_with_program_id;

pub fn close_position_orca_instruction(authority: &Pubkey, tuna_position: &TunaPosition, token_program_a: &Pubkey, token_program_b: &Pubkey) -> Instruction {
    let tuna_position_address = get_tuna_position_address(&tuna_position.position_mint).0;

    let ix_builder = ClosePositionOrca {
        authority: *authority,
        mint_a: Default::default(),
        mint_b: Default::default(),
        tuna_position: tuna_position_address,
        tuna_position_mint: tuna_position.position_mint,
        tuna_position_ata: get_associated_token_address_with_program_id(&tuna_position_address, &tuna_position.position_mint, &spl_token_2022::ID),
        tuna_position_ata_a: get_associated_token_address_with_program_id(&tuna_position_address, &tuna_position.mint_a, token_program_a),
        tuna_position_ata_b: get_associated_token_address_with_program_id(&tuna_position_address, &tuna_position.mint_b, token_program_b),
        whirlpool_program: orca_whirlpools_client::ID,
        orca_position: get_position_address(&tuna_position.position_mint).unwrap().0,
        token_program_a: *token_program_a,
        token_program_b: *token_program_b,
        token2022_program: spl_token_2022::ID,
    };

    ix_builder.instruction()
}
